bdns/dns.go

```diff
@@ -353,12 +379,28 @@ func (dnsClient *DNSClientImpl) exchangeOne(ctx context.Context, hostname string
 				hasRetriesLeft := tries < dnsClient.maxTries
 				if isRetryable && hasRetriesLeft {
 					tries++
-					// Chose a new server to retry the query with by incrementing the
-					// chosen server index modulo the number of servers. This ensures that
-					// if one dns server isn't available we retry with the next in the
-					// list.
-					chosenServerIndex = (chosenServerIndex + 1) % len(dnsClient.servers)
-					chosenServer = dnsClient.servers[chosenServerIndex]
+
+					if DNSServerHealthChecks && len(healthyServers) < 1 {
+						//too early to check if there's a healthy server
+						time.Sleep(periodDuration)
+					}
+					if DNSServerHealthChecks && len(healthyServers) >= 1 {
+						hsMX.RLock()
+						for k, v := range healthyServers {
```

Comment:
This changes the existing behavior, by making each retry randomly select from all of the healthy servers. I'd like to preserve the existing behavior, where we try to jump to the next server in the list (to avoid asking the same server the same question twice). But we would add something to that: If the next server on the list is unhealthy, continue skipping to the next server on the list until we find a healthy one *or* we have iterated enough times that we've looped through the whole list.